File: frontend/src/types/game.ts
// ABOUTME: TypeScript types for Game data matching backend Pydantic schemas
// ABOUTME: Ensures type safety between frontend and backend

import type { Player } from './player'

export interface Game {
  id: number
  player1_id: number
  player2_id: number
  winner_id: number
  created_at: string
  player1: Player
  player2: Player
  winner: Player
}

export interface GameCreate {
  player1_id: number
  player2_id: number
  winner_id: number
}

export interface GameListResponse {
  games: Game[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

File: backend/app/schemas/game.py
# ABOUTME: Pydantic schemas for game data validation and serialization
# ABOUTME: Defines request/response models for game API endpoints

from datetime import datetime

from pydantic import BaseModel, Field

from .player import PlayerResponse


class GameBase(BaseModel):
    """Base game schema with common fields"""

    player1_id: int = Field(..., description="First player ID")
    player2_id: int = Field(..., description="Second player ID")
    winner_id: int = Field(..., description="Winner player ID")


class GameCreate(GameBase):
    """Schema for creating a new game"""

    pass


class GameResponse(GameBase):
    """Schema for game response data"""

    id: int
    created_at: datetime
    player1: PlayerResponse
    player2: PlayerResponse
    winner: PlayerResponse

    class Config:
        from_attributes = True


class GameListResponse(BaseModel):
    """Schema for paginated game list response"""

    games: list[GameResponse]
    total: int
    page: int
    page_size: int
    total_pages: int
